netket/optimizer/rgn/rgn_pytree_logic.py

```diff
@@ -0,0 +1,358 @@
+    elif mode == "complex":
+        split_complex_params = True  # convert C→C and R&C→C to R→C
+        # centered_jacobian_fun = compose(stack_jacobian, centered_jacobian_cplx)
+
+        # avoid converting to complex and then back
+        # by passing around the oks as a tuple of two pytrees representing the real and imag parts
+        centered_jacobian_fun = compose(
+            stack_jacobian_tuple,
+            partial(centered_jacobian_cplx, _build_fn=lambda *x: x),
+        )
+        jacobian_fun = jacobian_cplx
+    elif mode == "holomorphic":
+        split_complex_params = False
+        centered_jacobian_fun = centered_jacobian_real_holo
+        jacobian_fun = jacobian_real_holo
+    else:
+        raise NotImplementedError(
+            'Differentiation mode should be one of "real", "complex", or "holomorphic", got {}'.format(
+                mode
+            )
+        )
+
+    if split_complex_params:
+        # doesn't do anything if the params are already real
+        params, reassemble = tree_to_real(params)
+
+        def f(W, σ):
+            return forward_fn(reassemble(W), σ)
+
+    else:
+        f = forward_fn
+
+    if pdf is None:
+        centered_oks = _divide_by_sqrt_n_samp(
+            centered_jacobian_fun(
+                f,
+                params,
+                samples,
+                chunk_size=chunk_size,
+            ),
+            samples,
+        )
+    else:
+        oks = jacobian_fun(f, params, samples)
+        oks_mean = jax.tree_map(partial(sum, axis=0), _multiply_by_pdf(oks, pdf))
+        centered_oks = jax.tree_multimap(lambda x, y: x - y, oks, oks_mean)
+
+        centered_oks = _multiply_by_pdf(centered_oks, jnp.sqrt(pdf))
+    if rescale_shift:
+        return _rescale(centered_oks)
+    else:
+        return centered_oks, None
+
+def avg_jacobian_real_holo(
+    forward_fn: Callable, params: PyTree, samples: Array, chunk_size: int = None
+) -> PyTree:
+
+    jac = jacobian_real_holo(forward_fn,params,samples,chunk_size)
+
+    return tree_mean(_divide_by_sqrt_n_samp(jac,samples))
+
+@partial(jax.jit, static_argnames=("forward_fn", "chunk_size"))
+def centered_rhessian_real_holo(
+    forward_fn: Callable, params: PyTree, samples: Array, connected_samples: Array, matrix_elements: Array, chunk_size: int = None
+) -> PyTree:
+
+    rhes = rhessian_real_holo(forward_fn,params,samples,connected_samples,matrix_elements,chunk_size)
+
+    return tree_subtract_mean(_divide_by_sqrt_n_samp(rhes,samples))
+
+def mat_vec(v: PyTree, oks: PyTree, rhes: PyTree, mean_grad: PyTree, eps: float, en: float, diag_shift: float) -> PyTree:
+    """
+    Compute ⟨O† O⟩v = ∑ₗ ⟨Oₖᴴ Oₗ⟩ vₗ
+    """
+    res = tree_conj(_vjp(oks, _jvp(rhes, v).conjugate()))
+    res2 = tree_conj(jax.tree_map(lambda x: _jvp(mean_grad, v)*x,v))
+    res3 = tree_conj(_vjp(oks, _jvp(oks, v).conjugate()))
+
+    res = jax.tree_multimap(lambda w,x,y,z: w - x - en*y + (1/eps)*(y + diag_shift*z),res,res2,res3,v)
```

Comment:
This does more matrix-vector products than needed. Also, the output vector is typically a lot smaller than the parameter vector, so we should rather sum what we can at that stage.
```suggestion
    # terms involving E_loc,j and the QGT
    w = _jvp(rhes, v) + (1/eps - en) * _jvp(oks, v)
    res = tree_conj(_vjp(oks, w.conjugate()))
    # g_i <O_j> term
    res2 = tree_conj(jax.tree_map(lambda x: _jvp(mean_grad, v)*x,v))

    res = jax.tree_multimap(lambda x,y,z: x - y + (diag_shift/eps) * z,res,res2,v)
```